src/mds/MDSRank.cc

```diff
@@ -3311,6 +3313,33 @@ void MDSRank::command_dump_inode(Formatter *f, const cmdmap_t &cmdmap, std::ostr
   }
 }
 
+void MDSRank::command_dump_dir(Formatter *f, const cmdmap_t &cmdmap, std::ostream &ss)
+{
+  std::lock_guard l(mds_lock);
+  std::string path;
+  bool got = cmd_getval(cmdmap, "path", path);
+  if (!got) {
+    ss << "missing path argument";
+    return;
+  }
+
+  CInode *in = mdcache->cache_traverse(filepath(path.c_str()));
+  if (!in) {
+    ss << "directory inode not in cache";
+    return;
+  }
+
+  f->open_array_section("dirs");
+  frag_vec_t leaves;
+  in->dirfragtree.get_leaves_under(frag_t(), leaves);
```

Comment:
Why didn't we directly use `get_dirfrags_under()`? And even if we use `get_leaves_under()`, shouldn't we check for emptiness of `leaves` prior to iterate over it?

Reply:
> Why didn't we directly use `get_dirfrags_under()`? And even if we use `get_leaves_under()`, shouldn't we check for emptiness of `leaves` prior to iterate over it?

Thanks! Here I refer to the code in `command_dirfrag_ls()` and there is a comment explaining why to use `get_leaves_under()`. An empty array will be returned if `leaves` is empty, or would it be better to replace it with an error message?

Reply:
This looks fine -- get_leaves_under() returns the entire set of frags.

Reply:
> > Why didn't we directly use `get_dirfrags_under()`? And even if we use `get_leaves_under()`, shouldn't we check for emptiness of `leaves` prior to iterate over it?
> 
> Thanks! Here I refer to the code in `command_dirfrag_ls()` and there is a comment explaining why to use `get_leaves_under()`. An empty array will be returned if `leaves` is empty, or would it be better to replace it with an error message?

Okay, I went through the code in `command_dirfrag_ls()`. Looks fine.